datalad/downloaders/tests/test_providers.py

```diff
@@ -85,4 +91,77 @@ def test_get_downloader_class():
     with patch.object(external_versions, '_versions', {'requests': None}):
         with assert_raises(RuntimeError) as cmr:
             Provider._get_downloader_class(url)
-        assert_in("you need 'requests'", str(cmr.exception))
\ No newline at end of file
+        assert_in("you need 'requests'", str(cmr.exception))
+
+@patch.multiple("appdirs.AppDirs", site_config_dir=None, user_config_dir=None)
+def test_Providers_from_config__files():
+    """Test configuration file precedence
+
+    Ensure that provider precedence works in the correct order:
+
+        datalad defaults < dataset defaults < system defaults < user defaults
+    """
+
+    # Setup. This would be nice to have a context manager for, but
+    # "with tempdir.TemporaryDirectory" wouldn't work on Python 2.7..
+
+    # Create a fake system directory, user directory, and dataset directory
+    sysdir = mkdtemp("sys", "dataladtest", None)
+    os.mkdir(sysdir + "/providers/", 0o700)
+    userdir = mkdtemp("user", "dataladtest", None)
+    os.mkdir(userdir + "/providers/", 0o700)
+    dsdir = mkdtemp("ds", "dataladtest", None)
+    os.makedirs(dsdir+ "/.datalad/providers/", 0o700)
+    # datalad looks for the .git directory to find the root of the
+    # dataset, not the .datalad directory.
+    os.makedirs(dsdir+ "/.git", 0o700)
+
+    with open(dsdir+"/.datalad/providers/atest.cfg", "w") as f:
+        f.write("""[provider:dscrcns]
+url_re = https?://crcns\.org/.*
+authentication_type = none
+        """)
+    with open(sysdir+"/providers/atest2.cfg", "w") as f:
+        f.write("""[provider:syscrcns]
+url_re = https?://crcns\.org/.*
+authentication_type = none
+        """)
+    with open(userdir+"/providers/atest3.cfg", "w") as f:
+        f.write("""[provider:usercrcns]
+url_re = https?://crcns\.org/.*
+authentication_type = none
+        """)
+
+    # Test the default, this is an arbitrary provider used from another
+    # test
+    providers = Providers.from_config_files(reload=True)
+    provider = providers.get_provider('https://crcns.org/data....')
+    assert_equal(provider.name, 'crcns')
+
+    # Test that the dataset provider overrides the datalad
+    # default
+    owd = os.getcwd()
+    os.chdir(dsdir)
```

Comment:
Just FYI: avoid direct `os.chdir`at all costs in datalad... use `datalad.utils.chpwd` if really really needed.

Here, use `with chpwd(dsdir):` context manager